### PR TITLE
New instrument: IPG Photonics YAR Amplifier series

### PR DESCRIPTION
**Pull request by bmoneke**
A new instrument definition.

### CODE REVIEW

File: pymeasure/instruments/ipgphotonics/yar.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+from pymeasure.instruments import Instrument, validators
+from pyvisa.constants import Parity, StopBits
+
+
+def emission_validator(value, values):
+    if value in values:
+        return value
+    elif isinstance(value, str):
+        if value.capitalize() in values:
+            return value.capitalize()
```
Comment (bilderbuchi) [inline pymeasure/instruments/ipgphotonics/yar.py:36]: I think the first clause is redundant, as it's fully covered by the second?

**Comment by bilderbuchi** [comment]
Not sure if you are already looking for a review, but I skimmed this and it looks good to me. (what else? :smile: )

**Comment by bmoneke** [comment]
You are invited to review.
I just did not want to pressure you to review.

From my side, the driver is ready for review.

File: pymeasure/instruments/ipgphotonics/yar.py (hunk 1 above)
Comment (bmoneke) [reply pymeasure/instruments/ipgphotonics/yar.py]: I streamlined the validator (first checks for boolean values, as that is the preferred way to interact).
With that change, I noticed a bug with the tests: it should be `upper`, not `capitalize`.

Now I test for a variety of input parameters.

File: pymeasure/instruments/ipgphotonics/yar.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+    emission = Instrument.control(
+        "STA",
+        "EM%s",
+        """Control emission of the amplifier.""",
+        cast=int,
+        values=("ON", "OFF"),
+        validator=emission_validator,
+        get_process=lambda v: bool(v & YAR.Status.EMISSION),
+        check_set_errors=True,
```
Comment (LongnoseRob) [inline pymeasure/instruments/ipgphotonics/yar.py:132]: Would it not make sense to map "ON" and "OFF" to `True` and `False` ?

**Review by LongnoseRob — APPROVED** [review]
Hello, 
looks like very much straight forward implementation.
I have only one comment on the `emission` propery,

And a more general question: does this device not have soem kind of interlock-circuit?

File: pymeasure/instruments/ipgphotonics/yar.py (hunk 2 above)
Comment (bmoneke) [reply pymeasure/instruments/ipgphotonics/yar.py:132]: `lambda v: bool(v & YAR.Status.EMISSION)`

does exactly that:
1. it reads the emission bit from the status byte (there is no dedicated "read emission command")
2. turn the bit to a boolean (True or False

On setting, I use the "emission_validator" to turn a boolean or "ON", "OFF" to the expected strings "ON" or "OFF".
So, you may use a boolean as well.

**Comment by bmoneke** [comment]
> And a more general question: does this device not have soem kind of interlock-circuit?

Yes it has it, but you cannot do anything via remote command with that interlock.

File: pymeasure/instruments/ipgphotonics/yar.py (hunk 2 above)
Comment (LongnoseRob) [reply pymeasure/instruments/ipgphotonics/yar.py:132]: Oh yes, I did not realize that.

Comment (bmoneke) [reply pymeasure/instruments/ipgphotonics/yar.py]: I improved the docstring, stating that it has a boolean value.

Comment (LongnoseRob) [reply pymeasure/instruments/ipgphotonics/yar.py]: LGTM

File: pymeasure/instruments/ipgphotonics/yar.py (hunk 1 above)
Comment (bilderbuchi) [reply pymeasure/instruments/ipgphotonics/yar.py]: thanks!

File: pymeasure/instruments/ipgphotonics/yar.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    @property
+    def status(self):
+        """Get the current status."""
+        got = int(self.values("STA")[0])
+        return self.Status(got)
+
+    emission = Instrument.control(
+        "STA",
+        "EM%s",
+        """Control emission of the amplifier (bool).""",
```
Comment (bilderbuchi) [inline pymeasure/instruments/ipgphotonics/yar.py:127]: In line with our guidelines, I think `emission_enabled` might be a better name for this boolean property.

File: tests/instruments/ipgphotonics/test_yar.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+def test_clear():
+    with expected_protocol(
+            YAR,
+            init_comm + [("RERR", "RERR")]
+    ) as inst:
+        inst.clear()
+
+
+def test_model():
+    """This test tests already getting min/max output power."""
```
Comment (bilderbuchi) [inline tests/instruments/ipgphotonics/test_yar.py:211]: docstring of this and the next test probably copy/pasted?

File: tests/instruments/ipgphotonics/test_yar.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+def test_emission_getter():
+    with expected_protocol(
+            YAR,
+            init_comm + [("STA", "STA:1")]
+    ) as inst:
+        assert inst.emission is True
+
+
+@pytest.mark.parametrize("value", (True, "ON", "on", "On", "oN"))
```
Comment (bilderbuchi) [inline tests/instruments/ipgphotonics/test_yar.py:71]: In my opinion, we should strive for conciseness/clarity/symmetry in our interfaces. I can see the attraction of being very "general" in what to accept (and in principle that's known [a good thing](https://en.wikipedia.org/wiki/Robustness_principle)). 
However, wouldn't it be preferable to just have `True`/`False` (and the equivalent truthy things that Python accepts) as the interface for this property? 
Accepting all kinds of variations of "on" could be a slippery slope towards [YAML](https://yaml.org/type/bool.html), which accepts  
``` 
y|Y|yes|Yes|YES|n|N|no|No|NO
|true|True|TRUE|false|False|FALSE
|on|On|ON|off|Off|OFF
```
Imo, this is not something we should mimick. Curious to hear your thoughts.

Comment (bmoneke) [reply tests/instruments/ipgphotonics/test_yar.py:71]: I agree: bool is sufficient.

Looking at other people's code I notice it, too, but changing my own code (predating my pymeasure time), I overlooked it... For that reason we have reviews.

File: tests/instruments/ipgphotonics/test_yar.py (hunk 4 above)
Comment (bmoneke) [reply tests/instruments/ipgphotonics/test_yar.py:211]: Yes, Done.

File: pymeasure/instruments/ipgphotonics/yar.py (hunk 3 above)
Comment (bmoneke) [reply pymeasure/instruments/ipgphotonics/yar.py:127]: Ok, done.

**Comment by bmoneke** [comment]
I resolved your issues, @bilderbuchi , can we merge?

**Comment by bilderbuchi** [comment]
Yes, thank you!

**Pull request merged**